server/events.go

```diff
@@ -0,0 +1,271 @@
+// Copyright 2018 The NATS Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"encoding/json"
+	"fmt"
+	"strconv"
+	"time"
+)
+
+const (
+	connectEventSubj    = "$SYS.%s.CLIENT.CONNECT"
+	disconnectEventSubj = "$SYS.%s.CLIENT.DISCONNECT"
+)
+
+// ConnectEventMsg is sent when a new connection is made that is part of an account.
+type ConnectEventMsg struct {
+	Server ServerInfo `json:"server"`
+	Client ClientInfo `json:"client"`
+}
+
+// DisconnectEventMsg is sent when a new connection previously defined from a
+// ConnectEventMsg is closed.
+type DisconnectEventMsg struct {
+	Server   ServerInfo `json:"server"`
+	Client   ClientInfo `json:"client"`
+	Sent     DataStats  `json:"sent"`
+	Received DataStats  `json:"received"`
+	Reason   string     `json:"reason"`
+}
+
+type ServerInfo struct {
+	Host string `json:"host"`
+	ID   string `json:"id"`
+	Seq  uint64 `json:"seq"`
+}
+
+// ClientInfo is detailed information about the client forming a connection.
+type ClientInfo struct {
+	Start   time.Time  `json:"start,omitempty"`
+	Host    string     `json:"host,omitempty"`
+	ID      uint64     `json:"id"`
+	Account string     `json:"acc"`
+	User    string     `json:"user,omitempty"`
+	Name    string     `json:"name,omitempty"`
+	Lang    string     `json:"lang,omitempty"`
+	Version string     `json:"ver,omitempty"`
+	Stop    *time.Time `json:"stop,omitempty"`
+}
+
+// DataStats reports how may msg and bytes. Applicable for both sent and received.
+type DataStats struct {
+	Msgs  int64 `json:"msgs"`
+	Bytes int64 `json:"bytes"`
+}
+
+// This will send a message.
+// TODO(dlc) - Note we want the sequence numbers to be serialized but right now they may not be.
+func (s *Server) sendInternalMsg(r *SublistResult, subj string, msg []byte) {
+	if s.sys == nil {
+		return
+	}
+	c := s.sys.client
+	acc := s.sys.account
+
+	// Prep internl structures needed to send message.
+	c.pa.subject = []byte(subj)
```

Comment:
It looks to be that `sendInternalMsg()` can be called concurrently, but it looks here that we use a single client object stored in server struct. So there should be theoretically several go routines changing the c.pa.subject/size/etc..? 
Did you mean to have this function be protected by some lock?

Reply:
Let me look into it. c.pa not setup to be shared as you know but you raise a good point.

Reply:
Yes, but in this case, we use `client` (which is s.sys.client) from various go routines, each time we want to produce a sys event. So this will be shared. Even the rest (calling processMsgResults, etc..) I think should be protected with a mutex.

Reply:
I understand. I had to do the pa.(..) stuff at the end to make it work, I think I have a comment somewhere in there that say something about coordination around outbound seq, etc. I could create a new mutex or spin up a Go routine that does the actual sending. I think I prefer the second option since we don't know all that is getting locked below us in process msg results.

Reply:
I personally would go with a dedicated lock, say in the `internal` structure. Spinning go routing to parse result, etc.. will cause out of order deliveries.

Reply:
(maybe you meant a single go-routine). Also were you referring to the situation of when dealing with an event we could end-up generating another kind of event in reaction of the first?

Reply:
Would just simulate readLoop like we have for normal clients.

Reply:
Yes just single Go routine.